src/cockpit/peer.py

```diff
@@ -19,125 +19,99 @@
 import logging
 import os
 
-from typing import Dict, List, Optional, Sequence
+from typing import Callable, Dict, List, Optional, Sequence
 
-from .router import Endpoint, Router, RoutingError, RoutingRule
+from .router import Endpoint, Router, RoutingRule
 from .protocol import CockpitProtocolClient
 from .transports import SubprocessTransport, SubprocessProtocol
 
 logger = logging.getLogger(__name__)
 
 
-class PeerStateListener:
-    def peer_state_changed(self, peer: 'Peer', event: str, exc: Optional[Exception] = None) -> None:
-        """Signal a state change in the Peer.
-
-        This is called on:
-            - connection made (event='connect')
-            - init message received (event='init')
-            - connection lost (event='closed', exc possibly set)
-        """
-
-    def peer_authorization_request(self, peer: 'Peer', message: Optional[str], prompt: str, echo: bool) -> None:
-        """Request authentication for connecting to the peer.
-
-        The state listener should respond by calling .authorize_response() on the Peer.
-        """
+class PeerError(Exception):
+    def __init__(self, problem, **kwargs):
+        self.problem = problem
+        self.kwargs = kwargs
 
 
 class Peer(CockpitProtocolClient, SubprocessProtocol, Endpoint):
     name: str
+    done_callbacks: List[Callable[[], None]]
     init_host: str
-    state_listener: Optional[PeerStateListener]
-
-    authorize_pending: Optional[str] = None  # the cookie of the pending request
 
-    def __init__(self,
-                 router: Router,
-                 name: str,
-                 state_listener: Optional[PeerStateListener] = None,
-                 init_host: Optional[str] = None):
+    def __init__(self, router: Router, name: str, init_host: Optional[str] = None):
         super().__init__(router)
 
         # All Peers start out frozen — we only unfreeze after we see the first 'init' message
         self.freeze_endpoint()
 
         self.name = name
-        self.state_listener = state_listener
+        self.done_callbacks = []
 
         assert router.init_host is not None
         self.init_host = init_host or router.init_host
 
-        self.authorize_pending = None
+    # Initialization
+    async def do_connect_transport(self) -> asyncio.Transport:
+        raise NotImplementedError
 
-    def spawn(self, argv: Sequence[str], env: Sequence[str], **kwargs) -> asyncio.Transport:
+    async def spawn(self, argv: Sequence[str], env: Sequence[str], **kwargs) -> asyncio.Transport:
+        # Not actually async...
```

Comment:
So why is it declared async then? Do subclasses need to override this function? (not in this commit)

Reply:
It's complicated.

By all rights, this should probably properly be async.  Compare with `asyncio.create_subprocess_exec()` for example.  If we ever move to using that function, then this would definitely need to be made async.

The funny thing is, though: even though asyncio's own subprocesses APIs are marked async, they're also just calling `subprocess.Popen()` inside, in the same blocking way that we are....

So, it's a bit of a shoulder shrug.  This should be async, but it isn't, and even if it called another async API, it would still not be async.  :shrug: